src/lib/io/include/sol-uart.h

```diff
@@ -81,22 +90,56 @@ enum sol_uart_parity {
  * @brief Amount of stop bits.
  */
 enum sol_uart_stop_bits {
-    SOL_UART_STOP_BITS_ONE = 0,
-    SOL_UART_STOP_BITS_TWO
+    SOL_UART_STOP_BITS_ONE = 0, /**< Use one stop bit*/
+    SOL_UART_STOP_BITS_TWO /**< Use two stop bits */
 };
 
+/**
+ * @struct sol_uart_config
+ * @brief A configuration struct used to set the UART paramenters.
+ *
+ * @see sol_uart_open()
+ */
 struct sol_uart_config {
 #ifndef SOL_NO_API_VERSION
-#define SOL_UART_CONFIG_API_VERSION (1)
-    uint16_t api_version;
+#define SOL_UART_CONFIG_API_VERSION (1) /**< compile time API version to be checked during runtime */
+    uint16_t api_version; /**< must match #SOL_UART_CONFIG_API_VERSION at runtime */
 #endif
-    enum sol_uart_baud_rate baud_rate;
-    enum sol_uart_data_bits data_bits;
-    enum sol_uart_parity parity;
-    enum sol_uart_stop_bits stop_bits;
-    void (*rx_cb)(void *user_data, struct sol_uart *uart, uint8_t byte_read); /** Set a callback to be called every time a character is received on UART */
-    const void *rx_cb_user_data;
-    bool flow_control; /** Enables software flow control(XOFF and XON) */
+    /**
+     * @brief Callback containing data was read from UART
+     *
+     * @param data User data as provided in sol_uart_config::user_data
+     * @param uart The UART handle
+     * @param buf The UART data
+     *
+     * @return The number of bytes read from @c buf or negative errno on error.
+     */
+    ssize_t (*rx_cb)(void *data, struct sol_uart *uart, struct sol_buffer *buf);
```

Comment:
or maybe `sol_str_slice` (you can mention it is a binary data and thus may contain inner \0 or no \0 at all, thus `len` must be considered)